File: internal/rtdx/generate.go
package rtdx

import (
	"strings"
	"time"

	"github.com/Karthik99999/pmd-gen/internal/romdata"
)

// rescueData contains info for a rescue/revival password
type rescueData struct {
	Timestamp int   `json:"timestamp"`
	Type      int   `json:"type"`
	Team      []int `json:"team"`
	Dungeon   int   `json:"dungeon"`
	Floor     int   `json:"floor"`
	Pokemon   int   `json:"pokemon"`
	Gender    int   `json:"gender"`
	Reward    int   `json:"reward"`
	Revive    int   `json:"revive"`
}

// NewRescueData creates a new instance of RescueData
func NewRescueData(pswdType int, teamName string, info ...int) *rescueData {
	data := &rescueData{}
	data.Timestamp = int(time.Now().Unix())
	data.Type = pswdType
	var team []int
	for i := 0; i < 12; i++ {
		if i < len(teamName) {
			team = append(team, strings.Index(romdata.GetRomData().CharmapText, string(teamName[i]))-394)
			// idk why I have to subtract 394 but it works
		} else {
			team = append(team, 0)
		}
	}
	data.Team = team

	if pswdType == 0 {
		data.Dungeon = info[0]
		data.Floor = info[1]
		data.Pokemon = info[2]
		data.Gender = info[3]
		data.Reward = info[4]
	} else {
		data.Revive = info[0]
	}

	return data
}

// shuffle password symbols
func shuffle(code []string) []string {
	unshuffledIndex := []int{3, 27, 13, 21, 12, 9, 7, 4, 6, 17, 19, 16, 28, 29, 23, 20, 11, 0, 1, 22, 24, 14, 8, 2, 15, 25, 10, 5, 18, 26}
	shuffled := make([]string, 30)
	for i := 0; i < len(unshuffledIndex); i++ {
		shuffled[unshuffledIndex[i]] = code[i]
	}
	return shuffled
}

// convert symbols to indexes 0-63
// xs is not used in passwords due to 64 being a 7 bit number
func toSymbols(bitstream []int) []string {
	symbols := []string{"1f", "2f", "3f", "4f", "5f", "6f", "7f", "8f", "9f", "pf", "mf", "df", "xf",
		"1h", "2h", "3h", "4h", "5h", "6h", "7h", "8h", "9h", "ph", "mh", "dh", "xh",
		"1w", "2w", "3w", "4w", "5w", "6w", "7w", "8w", "9w", "pw", "mw", "dw", "xw",
		"1e", "2e", "3e", "4e", "5e", "6e", "7e", "8e", "9e", "pe", "me", "de", "xe",
		"1s", "2s", "3s", "4s", "5s", "6s", "7s", "8s", "9s", "ps", "ms", "ds"}
	code := make([]string, 30)
	for i := 0; i < len(code); i++ {
		code[i] = symbols[bitstream[i]]
	}
	return code
}

// bitpack the code
func bitunpack(bytearr []int) []int {
	var bitstream []int
	reader := NewBitstreamReader(bytearr, 8)
	for reader.Remaining() {
		bitstream = append(bitstream, reader.Read(6))
	}
	return bitstream
}

// encrypt the code using rng
func encrypt(data []int) []int {
	encrypted := []int{data[0], data[1]}
	seed := data[0] | data[1]<<8
	rng := NewRNG(seed)
	for i := 2; i < len(data); i++ {
		rand := rng.NextInt()
		encrypted = append(encrypted, (data[i]+rand)&0xFF)
	}
	remain := 8 - (len(data) * 8 % 6)
	encrypted[len(encrypted)-1] &= (1 << remain) - 1
	return encrypted
}

func (rd *rescueData) Serialize() []string {
	writer := NewBitstreamWriter(8)
	writer.Write(rd.Timestamp, 32)
	writer.Write(rd.Type, 1)
	writer.Write(0, 1) // Unknown, Can be either 0 or 1
	for i := 0; i < 12; i++ {
		if i < len(rd.Team) {
			writer.Write(rd.Team[i], 9)
		} else {
			writer.Write(0, 9)
		}
	}
	if rd.Type == 0 {
		writer.Write(rd.Dungeon, 7)
		writer.Write(rd.Floor, 7)
		writer.Write(rd.Pokemon, 11)
		writer.Write(rd.Gender, 2)
		writer.Write(rd.Reward, 2)
		writer.Write(0, 1) // Unknown, Can be either 0 or 1
	} else {
		writer.Write(rd.Revive, 30)
	}

	data := writer.Finish()
	data = append([]int{checksum(data)}, data...)
	// Serealize data
	encrypted := encrypt(data)
	bitstream := bitunpack(encrypted)
	symbols := toSymbols(bitstream)
	code := shuffle(symbols)
	return code
}

File: src/lib/generators/rtdx/generate.ts
import { bitpack, BitstreamWriter } from '../bitstream';
import { symbols, checksum, RNG } from './utils';
import { deserialize } from './read';
import Data from './data';

function shuffle(code: string[]): string[] {
	const shuffledIndexes = [
		3, 27, 13, 21, 12, 9, 7, 4, 6, 17, 19, 16, 28, 29, 23, 20, 11, 0, 1, 22, 24, 14, 8, 2, 15, 25, 10, 5, 18, 26,
	];
	const shuffled = [];
	for (let i = 0; i < 30; i++) {
		shuffled[shuffledIndexes[i]] = code[i];
	}
	return shuffled;
}

function bitsToSymbols(bits: number[]): string[] {
	return bits.map((i) => symbols[i]);
}

function encrypt(code: number[]): number[] {
	const newcode = code.slice(0, 2);
	const seed = code[0] | (code[1] << 8);
	const rng = new RNG(seed);
	for (const c of code.slice(2)) {
		const rand = rng.next();
		newcode.push((c + rand) & 0xff);
	}
	const remain = 8 - ((code.length * 8) % 6);
	newcode[newcode.length - 1] &= (1 << remain) - 1;
	return newcode;
}

interface PasswordData {
	timestamp: number;
	team: number[];
	dungeon: number;
	floor: number;
	pokemon: number;
	/**
	 * 1 = Regular,
	 * 2 = Special,
	 * 3 = Deluxe
	 */
	reward: 1 | 2 | 3;
	revive: number;
}

type RescueData = Omit<PasswordData, 'revive'> & { type: 0 };
type RevivalData = Omit<PasswordData, 'dungeon' | 'floor' | 'pokemon' | 'gender' | 'reward'> & { type: 1 };

function serialize(data: RescueData | RevivalData): string {
	const writer = new BitstreamWriter(8);
	writer.write(data.timestamp, 32);
	writer.write(data.type, 1);
	writer.write(0, 1); // Unknown, Can be either 0 or 1
	for (let i = 0; i < 12; i++) {
		if (i < data.team.length) {
			writer.write(data.team[i], 9);
		} else {
			writer.write(0, 9);
		}
	}
	if (data.type === 0) {
		writer.write(data.dungeon, 7);
		writer.write(data.floor, 7);
		writer.write(data.pokemon, 11);
		writer.write(0, 2); // Gender
		writer.write(data.reward, 2);
		writer.write(0, 1); // Unknown, Can be either 0 or 1
	} else {
		writer.write(data.revive, 30);
	}

	const code = writer.finish();
	code.unshift(checksum(code));
	const encrypted = encrypt(code);
	const bitstream = bitpack(encrypted, 8, 6);
	const bits = bitsToSymbols(bitstream);
	const shuffled = shuffle(bits);
	return shuffled.join('');
}

export function generateRescue(data: { team: string; dungeon: number; floor: number; pokemon: number }): string {
	if (data.team.length < 1 || data.team.length > 12) throw new Error('Team name must be between 1 and 12 characters');
	const team: number[] = [];
	for (const char of data.team) {
		const index = Data.charmap_text.indexOf(char);
		if (index < 0) {
			throw new Error(`Invalid character in team name: ${char}`);
		}
		team.push(index);
	}
	const dungeonData = Data.dungeons.get(data.dungeon);
	if (!dungeonData.valid) {
		throw new Error('Invalid dungeon');
	}
	if (data.floor < 1 || data.floor > dungeonData.floors) {
		throw new Error('Invalid floor');
	}

	if (!Data.pokemon.get(data.pokemon).valid) {
		throw new Error('Invalid Pokemon');
	}

	const rescueData: RescueData = {
		timestamp: Math.round(Date.now() / 1000),
		type: 0,
		team,
		dungeon: data.dungeon,
		floor: data.floor,
		pokemon: data.pokemon,
		reward: 3,
	};
	return serialize(rescueData);
}

export function generateRevival(password: string, teamName: string): string {
	const data = deserialize(password);
	if (data.inclChecksum !== data.calcChecksum) throw new Error('The provided rescue password was invalid');

	if (teamName.length < 1 || teamName.length > 12) throw new Error('Team name must be between 1 and 12 characters');
	const team: number[] = [];
	for (const char of teamName) {
		const index = Data.charmap_text.indexOf(char);
		if (index < 0) {
			throw new Error(`Invalid character in team name: ${char}`);
		}
		team.push(index);
	}

	const revivalData: RevivalData = {
		timestamp: Math.round(Date.now() / 1000),
		type: 1,
		team,
		revive: data.revive,
	};
	return serialize(revivalData);
}
